stdlib/public/Atomics/UnsafeAtomicLazyReference.swift

```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+/// An atomic optional strong reference that can be set (initialized) exactly
+/// once, but read many times.
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+@frozen
+public struct UnsafeAtomicLazyReference<Instance: AnyObject> {
+  public typealias Value = Instance?
+  public typealias AtomicStorage = Instance?
+
+  @usableFromInline
+  internal let _ptr: UnsafeMutablePointer<AtomicStorage>
+
+  @_transparent // Debug performance
+  public init(@_nonEphemeral at pointer: UnsafeMutablePointer<AtomicStorage>) {
+    self._ptr = pointer
+  }
+}
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension UnsafeAtomicLazyReference {
+  @inlinable
+  public static func create() -> Self {
+    let ptr = UnsafeMutablePointer<AtomicStorage>.allocate(capacity: 1)
+    ptr.initialize(to: nil)
+    return Self(at: ptr)
+  }
+
+  public func destroy() {
```

Comment:
Should you release a non-nil unmanaged here?

Reply:
No -- unmanaged references aren't necessarily owning (a.k.a +1). If they are, they need to manually released before deallocation.

Reply:
Now I see how it happens. Don't try this at home!

Reply:
The latest revision switches atomic lazy references to use trivial storage representation, so it is now critical to call `dispose` before deinitializing storage.